src/relations/db.py

```diff
@@ -0,0 +1,414 @@
+                "state": self._get_state(dbconnstr),
+            }
+            databag.update(updates)
+
+    def generate_username(self, event):
+        """Generates a username for this relation."""
+        return f"relation_id_{event.relation.id}"
+
+    def _get_read_only_endpoint(self):
+        """Get a read-only-endpoint from backend relation.
+
+        Though multiple readonly endpoints can be provided by the new backend relation, only one
+        can be consumed by this legacy relation.
+        """
+        read_only_endpoints = self.charm.backend_relation_app_databag.get("read-only-endpoints")
+        if read_only_endpoints is None or len(read_only_endpoints) == 0:
+            return None
+        return read_only_endpoints.split(",")[0]
+
+    def _get_state(self, standbys: str) -> str:
+        """Gets the given state for this unit.
+
+        Args:
+            standbys: the comma-separated list of postgres standbys
+
+        Returns:
+            The described state of this unit. Can be 'standalone', 'master', or 'standby'.
+        """
+        if standbys == "":
+            return "standalone"
+        # TODO this doesn't ever return false. Revisit mastery once scaling is sorted, and check
+        # replicas return standby.
+        elif self.charm.unit.is_leader():
+            return "master"
+        else:
+            return "standby"
+
+    def _on_relation_departed(self, departed_event: RelationDepartedEvent):
+        """Handle db-relation-departed event.
+
+        Removes relevant information from pgbouncer config when db relation is removed. This
+        function assumes that relation databags are destroyed when the relation itself is removed.
+
+        This doesn't delete users or tables, following the design of the legacy charm.
+        """
+        logger.info("db relation removed - updating config")
+        logger.warning(
+            f"DEPRECATION WARNING - {self.relation_name} is a legacy relation, and will be deprecated in a future release. "
+        )
+
+        app_databag = departed_event.relation.data[self.charm.app]
+        unit_databag = departed_event.relation.data[self.charm.unit]
+
+        for databag in [app_databag, unit_databag]:
+            databag["allowed-units"] = self.get_allowed_units(departed_event.relation)
+
+    def _on_relation_broken(self, broken_event: RelationBrokenEvent):
+        """Handle db-relation-broken event.
+
+        Removes all traces of the given application from the pgbouncer config.
+        """
+        app_databag = broken_event.relation.data[self.charm.app]
+
+        cfg = self.charm.read_pgb_config()
+        user = app_databag["user"]
+        database = app_databag["database"]
+
+        # delete user
+        self.charm.remove_user(user, cfg=cfg, render_cfg=True, reload_pgbouncer=True)
+        try:
+            logging.error(self.charm.backend_postgres.delete_user)
+            self.charm.backend_postgres.delete_user(user, if_exists=True)
```

Comment:
I believe we need to add a check to ensure the backend PostgreSQL relation is still active (otherwise PostgreSQL would have removed the pgbouncer user and this can fail; it will be handled also in some way in the PostgreSQL side later, to delete users that PgBouncer created if the relation is broken between them).

Reply:
This is a good idea, I'll implement this.

Reply:
So this was a little more complicated than I thought, but I've added two checks: one checks at the beginning of the hook that the backend and db relations haven't been initialised, so if this hook somehow gets called before relation_joined (due to the deferral queue getting a bit weird), we should wait for the relation to initialise, then immediately remove. I'd love to just remove the relation entirely, but since we can't control the hook queue I think this is the best I can do. 

The second check verifies backend_postgres when it's used, when we're deleting user. If it's inaccessible at this point, we assume the backend relation has been removed, and postgres will remove the users pgbouncer has created, so we skip deleting the user. Every method that deletes users shouldn't raise errors if the user in question isn't deleted, so operation should continue even if the users are being deleted in some other way. 

Does this work for you?

Reply:
Yeah It works! Thanks Will!